rclpy/src/rclpy/_rclpy_qos_event.c

```diff
@@ -30,14 +33,36 @@ static
 bool
 _check_rcl_return(rcl_ret_t ret, const char * error_msg)
 {
-  if (RCL_RET_OK != ret) {
-    PyErr_Format(
-      PyExc_RuntimeError,
-      "%s: %s", error_msg, rcl_get_error_string().str);
-    rcl_reset_error();
-    return false;
+  if (RCL_RET_OK == ret) {
+    return true;
+  }
+
+  PyObject * exception = PyExc_RuntimeError;
+  PyObject * pyqos_event_module = NULL;
+  PyObject * pyqos_event_exception = NULL;
+
+  if (RCL_RET_UNSUPPORTED == ret) {
+    PyObject * pyqos_event_module = PyImport_ImportModule("rclpy.qos_event");
```

Comment:
I think that it's better to define the new exception in `C`, like this:
https://github.com/ros2/rclpy/blob/3d1698a99cbde82322c82fde428c76dd2242e855/rclpy/src/rclpy/_rclpy.c#L5523.

Reply:
If we define the new exception in C, then it looks like using it from Python module would be something like:
```
from rclpy.impl.implementation_singleton import rclpy_implementation as _rclpy
...
try:
    ...
except _rclpy.UnsupportedEventTypeError as e:
    ...
```
whereas if we define it in Python (as it is in this pull request currently), the usage would be like:
```
from rclpy.qos_event import UnsupportedEventTypeError
...
try:
    ...
except UnsupportedEventTypeError as e:
    ...
```
I think the latter is better, because you probably wouldn't want a user to be importing `rclpy.impl.*`

Reply:
you can import it in `rclpy.qos_event` and then reexport it.

@wjwwood for a second opinion.

Reply:
I think both will work. It’s slightly less awkward to avoid importing code in the c extension and then reexport it as @ivanpauno said, but I don’t mind as long as it works.

Reply:
I have moved the definition of the new exception into C instead of Python.